### PR TITLE
Track synthetic source for disabled objects

### PR DESCRIPTION
**Pull request by kkrik-es**
Tracks source for disabled object in `_ignored_source` metadata field.

Related to #106825

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java [hunk 1]
```diff
@@ -124,6 +124,15 @@ private static void internalParseDocument(MetadataFieldMapper[] metadataFieldsMa
...
+                if (context.isSourceSynthetic()) {
+                    context.addIgnoredField(
+                        new IgnoredSourceFieldMapper.NameValue(
+                            MapperService.SINGLE_MAPPING_NAME,
+                            0,
+                            XContentDataHelper.encodeToken(context.parser())
+                        )
+                    );
+                }
                 context.parser().skipChildren();
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:136]: should `context.parser().skipChildren()` only be executed if `context.isSourceSynthetic() == false`? If so then add an `else` clause here?

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java [hunk 2]
```diff
@@ -246,18 +255,28 @@ static Mapping createDynamicUpdate(DocumentParserContext context) {
...
+            if (context.isSourceSynthetic()) {
+                context.addIgnoredField(
+                    new IgnoredSourceFieldMapper.NameValue(
+                        context.parent().fullPath(),
+                        context.parent().fullPath().indexOf(currentFieldName),
+                        XContentDataHelper.encodeToken(parser)
+                    )
+                );
+            }
+            parser.skipChildren();
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:271]: same question as above.

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java [hunk 3]
```diff
@@ -637,6 +637,13 @@ public DocumentDimensions getDimensions() {
         return dimensions;
     }
 
+    /**
+     * Indicates if the index is configured to use synthetic source.
+     */
+    public boolean isSourceSynthetic() {
+        return indexSettings().getMode().isSyntheticSourceEnabled() || mappingLookup.isSourceSynthetic();
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java:644]: I wonder whether we can omit the `indexSettings().getMode().isSyntheticSourceEnabled()` part here? If `mappingLookup.isSourceSynthetic()` is `true` then `indexSettings().getMode().isSyntheticSourceEnabled()` should be evaluate `true`?

**Review by martijnvg — COMMENTED** [review]
Looks good. Left two questions.

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java (hunk 1 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:136]: I think it's fine as is, the `_ignored_source` entry created above includes all children so there's no point to further inspect this object. The tests support that this is the case?

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java (hunk 3 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java]: Good point, this is not needed.

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java (hunk 1 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:136]: I see, this is a no-op for the new path.

File: server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperTests.java [hunk 4]
```diff
@@ -145,4 +145,139 @@ public void testMultipleIgnoredFieldsManyObjects() throws IOException {
...
+                b.endObject();
+            }
+            b.endObject();
+        });
+        assertThat(syntheticSource, Matchers.containsString("\"boolean_value\":" + booleanValue));
+        assertThat(syntheticSource, Matchers.containsString("\"path\":{\"to\":{\"int_value\":" + intValue));
+        assertThat(syntheticSource, Matchers.containsString("\"some\":{\"deeply\":{\"nested\":{\"string_value\":\"" + stringValue + "\""));
+    }
+
+    public void testDisabledObjectSingleField() throws IOException {
```
[the remark this answers is not in the document]
Comment (kkrik-es) [reply server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperTests.java:196]: Done.

**Review by martijnvg — APPROVED** [review]
lgtm 👍

File: server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkrik-es) [reply server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperTests.java:196]: And a yaml test too.

**Review by lkts — APPROVED** [review]

**Comment by felixbarny** [comment]
Does this also track the source of objects where dynamic mapping is disabled (`dynamic: false`)?

**Comment by kkrik-es** [comment]
> Does this also track the source of objects where dynamic mapping is disabled (`dynamic: false`)?

No, but I'll take a look at that too.

**Pull request merged**